presto-hive/src/test/java/com/facebook/presto/hive/statistics/TestQuickStatsProvider.java

```diff
@@ -0,0 +1,341 @@
+                ImmutableList.of(quickStatsBuilderMock));
+
+        // Execute
+        ImmutableList<String> testPartitions1 = ImmutableList.of("partition1", "partition2", "partition3");
+        Map<String, PartitionStatistics> quickStats = quickStatsProvider.getQuickStats(SESSION, metastoreMock,
+                new SchemaTableName(TEST_SCHEMA, TEST_TABLE), metastoreContext, testPartitions1);
+
+        // Verify only one call was made for each test partition
+        assertEquals(quickStats.entrySet().size(), testPartitions1.size());
+        assertTrue(quickStats.keySet().containsAll(testPartitions1));
+        quickStats.values().forEach(ps -> assertEquals(ps, expectedPartitionStats));
+        for (String testPartition : testPartitions1) {
+            verify(quickStatsBuilderMock).buildQuickStats(any(), any(), any(), any(), eq(testPartition), any());
+        }
+        verifyNoMoreInteractions(quickStatsBuilderMock);
+
+        // For subsequent calls for the same partitions that are already cached, no new calls are mode to the quick stats builder
+        quickStatsProvider.getQuickStats(SESSION, metastoreMock,
+                new SchemaTableName(TEST_SCHEMA, TEST_TABLE), metastoreContext, testPartitions1);
+        verifyNoMoreInteractions(quickStatsBuilderMock);
+
+        // For subsequent calls with a mix of old and new partitions, we only see calls to the quick stats builder for the new partitions
+        ImmutableList<String> testPartitions2 = ImmutableList.of("partition4", "partition5", "partition6");
+        ImmutableList<String> testPartitionsMix = ImmutableList.<String>builder().addAll(testPartitions1).addAll(testPartitions2).build();
+
+        quickStats = quickStatsProvider.getQuickStats(SESSION, metastoreMock,
+                new SchemaTableName(TEST_SCHEMA, TEST_TABLE), metastoreContext, testPartitionsMix);
+        assertEquals(quickStats.entrySet().size(), testPartitionsMix.size());
+        assertTrue(quickStats.keySet().containsAll(testPartitionsMix));
+        quickStats.values().forEach(ps -> assertEquals(ps, expectedPartitionStats));
+        for (String testPartition : testPartitions2) {
+            verify(quickStatsBuilderMock).buildQuickStats(any(), any(), any(), any(), eq(testPartition), any());
+        }
+        verifyNoMoreInteractions(quickStatsBuilderMock);
+    }
+
+    @Test(invocationCount = 3)
+    public void testConcurrentFetchForSamePartition()
+            throws ExecutionException, InterruptedException
+    {
+        QuickStatsBuilder longRunningQuickStatsBuilderMock = mock(QuickStatsBuilder.class);
+        when(longRunningQuickStatsBuilderMock.buildQuickStats(any(), any(), any(), any(), any(), any()))
+                .thenAnswer((Answer<PartitionQuickStats>) invocationOnMock -> {
+                    // Sleep for 1.5s to simulate a long-running quick stats call
+                    Thread.sleep(1500);
```

Comment:
Similar vein to other comment about CI time. I would be prudent about forcing tests to wait for any amount of time. Would prefer to reduce timeouts or find a way to better simulate long-running calls. Don't have to make any change but would be nice to make this as small as possible without jeopardizing the stability of the test.

Reply:
reducing to 50ms